(SIMP-9023) Adding .gitlab-ci.yml

billy-davies opened the pull request:
The Travis CI testing was removed, and the GLCI should
preserve this functionality.

[SIMP-9023] #close

[SIMP-9023]: https://simp-project.atlassian.net/browse/SIMP-9023

billy-davies committed:
(SIMP-9023) Migrate Travis CI tests

Migrating Travis CI tests into GitLab CI

SIMP-9023 #close

billy-davies committed:
Updating the .gitlab-ci.yml to add rubocop tests.

billy-davies committed:
Merge branch 'master' into SIMP-9023

op-ct commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,141 @@
+#    - 'export PROJECT_DIR="set_travis_env_vars"'
+#    - 'rubocop --fail-level warning --display-only-fail-level-offenses $PROJECT_DIR'
+
+.lint_tests: &lint_tests
+  stage: 'validation'
+  tags: ['docker']
+  <<: *rubocop_project_dir
+  script:
+    - 'export PROJECT_DIR="octokit_repo_puncher"'
+    - 'rubocop --fail-level warning --display-only-fail-level-offenses $PROJECT_DIR'
```
This step won't install the project's `Gemfile` dependencies if it uses the keys defined in `&rubocop_project_dir` instead of `&setup_bundler_env`.

This probably wasn't clear from reading the original Travis pipeline—in Travis, bundle executes automatically before running `ruby` builds.

op-ct commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,141 @@
+  <<: *setup_bundler_env
+  script:
+    - bash -c 'for json in $(find . -type f -name "*.json"); do echo "== $json" && jq . "$json"; done'
+.lint_tests: &lint_tests
+  stage: 'validation'
+  tags: ['docker']
+  <<: *setup_bundler_env
+  script:
+    - bash -c 'find . -type f -name "*.sh"  | xargs shellcheck'
+
+.lint_tests: &lint_tests
+  stage: 'validation'
+  tags: ['docker']
+  <<: *setup_bundler_env
+  script:
+    - bash -c 'find . -type f -name "*.yaml" -o -name "*.yml" | xargs yaml-lint'
+
+.lint_tests: &lint_tests
+  stage: 'validation'
+  tags: ['docker']
+  <<: *setup_bolt_env
+  script:
+    - 'BOLT_DISABLE_ANALYTICS=yes bolt task show | grep releng::'
```
There are a lot of hidden `.lint_tests`keys and anchors

billy-davies replied:
Oops, bad copy paste, I meant to rename each job to their own thing.

op-ct commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,141 @@
+    # install yaml-lint for the linting phase:
+    - 'gem install --conservative --minimal-deps yaml-lint'
+
+.rubocop_project_dir: &rubocop_project_dir
+  cache:
+    untracked: true
+  before_script:
+    # install rubocop:
+    - 'gem install --conservative --minimal-deps rubocop'
```
It seems premature to give rubocop its own YAML anchor+hidden key; there's only one job in the pipeline uses it.  Instead, I'd move the `gem install --conservative --minimal-deps rubocop` line into the early steps of that job.

Also: beware of re-defining the `cache` and `before_script` key: a single job definition can only define one of each—this anchor and `&setup_bundler_env` couldn't be used in the same step without conflicting/overwriting these keys.

billy-davies replied:
Noted, changed the `&setup_bundler_env` to install rubocop as well and then just used that instead.

billy-davies committed:
Update .gitlab-ci.yml

Updated .gitlab-ci.yml after code review

billy-davies committed:
(SIMP-9023) Updating .gitlab-ci.yml

Updating .gitlab-ci.yml by tidying up the
variables list, and removing extra steps
in the setup_bundler_env phase.

op-ct commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,107 @@
+    untracked: true
+  before_script:
+    # install bolt:
+    - 'wget https://apt.puppet.com/puppet-tools-release-xenial.deb'
+    - 'sudo dpkg -i puppet-tools-release-xenial.deb'
+    - 'sudo apt-get update'
+    - 'sudo apt-get install puppet-bolt'
+    - 'command -v bolt || if command -v apt-get; then apt-get update; apt-get install -y bolt; fi ||:'
```
These `sudo` commands cause [this failure](https://gitlab.com/simp/releng-misc/-/jobs/985598129#L37) inside the `bolt_lint` job

billy-davies replied:
Removed sudo from the issued commands

op-ct commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,107 @@
+    # install rubocop:
+    - 'gem install --conservative --minimal-deps rubocop'
+
+.setup_bolt_env: &setup_bolt_env
```
This YAML anchor is only used by a single job, `bolt_lint`, so it's probably more complex and fragile than its worth to keep these here.  I'd suggest moving these `before_script` steps into the `script` of the **`bolt_lint`** job.

billy-davies replied:
ok, sent it through

op-ct reviewed (changes requested):
Some GLCI pipeline steps fail with `ERROR: Preparation failed: no Docker image specified to run the build in`, because docker jobs need an [`image`](https://docs.gitlab.com/ee/ci/yaml/#image) (see [example from our modules](https://github.com/simp/pupmod-simp-vsftpd/blob/master/.gitlab-ci.yml#L219))

op-ct committed:
Merge branch 'master' into SIMP-9023

billy-davies committed:
Update .gitlab-ci.yml

removing the sudo commands to allow jobs to succeed. moving bolt setup anchored before phase to bolt lint script portion.

billy-davies committed:
Update .gitlab-ci.yml

Setting the default image to ruby:2.6 for all the unit tests.

op-ct commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,104 @@
+    - 'bundle show sync || :'
+    - 'bundle exec gem list sync || :'
+
+    # install yaml-lint for the linting phase:
+    - 'gem install --conservative --minimal-deps yaml-lint'
+
+    # install rubocop:
+    - 'gem install --conservative --minimal-deps rubocop'
```
These steps should be moved under the `script` section of the stages that need them

op-ct committed:
Fix some things, force Boltdir perms

op-ct committed:
Allow shellcheck to fail, remove bundler

op-ct commented:
@billy-davies I pushed a few fixes, notably allowing shellcheck to fail (as the Travis pipeline did), ensuring that the gitlab.com runners' umask doesn't cause the bolt tests to fail, and gating the apt/dpkg commands behind checks to keep them viable for non-debian runners.

The Ruby update bumped the rubocop definitions, which introduced new cops that are failing.  Until we get a chance to look at them, they should probably be allowed to fail like the shellchecks.

billy-davies committed:
Update .gitlab-ci.yml

Allowing rubocop jobs to fail.

op-ct committed:
Merge branch 'master' into SIMP-9023

op-ct committed:
Update .gitlab-ci.yml post-Boltdir
